syntax = "proto3";

package dfs_service;


service DFSService {


    rpc StoreFile(stream StoreRequest) returns (StoreResponse);
    rpc FetchFile(FetchRequest) returns (stream FetchResponse);
    rpc ListFiles(FileListRequest) returns (FileList);
    rpc GetFileStatus(StatusRequest) returns (FileInfo);
    rpc RequestWriteLock(WriteLockRequest) returns (WriteLockResponse);
    rpc CallbackList(FileListRequest) returns (FileList);
    rpc DeleteFile(DeleteRequest) returns (DeleteResponse);

}

message StoreRequest {
    string file_name = 1;
    bytes chunk = 2;
    string client_id = 3;
    uint32 crc = 4;
    int64 mtime = 5;

}

message StoreResponse {
    bool success = 1;
    string file_name = 2;
    uint32 crc = 3;
    bool file_lock = 5;
    string client_id = 6;
}


message FetchRequest {
    string file_name = 1;
    string client_id = 2;
    uint32 crc = 3;
    
}

message FetchResponse {
    bytes chunk = 1;
    string file_name = 2;
    uint32 crc = 3;
    string client_id = 4;

}

message FileListRequest {
    string client_id = 1;
    string name= 2;
}

message FileList {
    repeated FileInfo files = 1;
}

message DeleteRequest {
    string file_name = 1;
    string client_id = 2;
}

message DeleteResponse {
    bool success = 1;
}

message StatusRequest {
    string file_name = 1;
    string client_id = 2;       
}  

message FileInfo {
    string name = 1;
    int64 size = 2;
    int64 mtime = 3;
    int64 ctime = 4;
    uint32 crc = 5;
}

message WriteLockRequest {
    string filename = 1;
    string client_id = 2;
}

message WriteLockResponse {
    bool success = 1;
    string message = 2;
}

// Add your message types here

